[FSDP] Add `_set_flattened()`; `_is_flattened()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85038

For both exposing the original parameters and for TP integration, we cannot only rely on `isinstance(param, FlatParameter)` to ignore already-flattened parameters in `.named_parameters()`. As a simple workaround, we can mark original parameters or `ShardedTensor`s with an attribute `_fsdp_flattened` (saved as a string variable `FSDP_FLATTENED`) to indicate that the parameter/tensor has already been flattened. This issue only arises for recursive/nested FSDP wrapping.

This PR also changes `isinstance(param, FlatParameter)` checks to `type(param) is FlatParameter` because all tensor subclasses that have `_is_param == True` will return `True` for `isinstance(param, <any subclass with _is_param == True>)`. This means that a `ShardedTensor` parameter will return `True` for `isinstance(st, FlatParameter)`, which is not what we want.
https://github.com/pytorch/pytorch/blob/5271494ef21ae0140755a41f3b16a8bd745642b6/torch/nn/parameter.py#L8-L10

Commit:
[FSDP] Add `_set_flattened()`; `_is_flattened()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `_set_flattened()`; `_is_flattened()`"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -310,7 +311,7 @@ def _init_flat_param(
                     )
                 else:
-                    if isinstance(param, FlatParameter):
+                    if type(param) is FlatParameter:
```
does this mean that previously, any Parameter subclass with _is_param == True would mistakenly pass this check as a FlatParameter?

awgu replied:
Yes.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1152,7 +1153,7 @@ def _get_ignored_params(
             for m in ignored_modules
             for p in m.parameters()
-            if not isinstance(p, FlatParameter)
+            if not _is_flattened(p)
```
maybe rename to `_is_fsdp_flattened` or do `import _is_flattened as _is_fsdp_flattened` to provide clarity in case there are other types of flatenning going on in the future.

awgu replied:
Sounds good. I can rename it to `_is_fsdp_flattened()`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1457,7 +1458,7 @@ def _get_orig_params(
             while True:
                 param = next(param_gen)
-                if param not in ignored_params and not isinstance(param, FlatParameter):
+                if param not in ignored_params and not _is_flattened(param):
```
I'm wondering why we need `_is_flattened`, can't we just check type(param) == FlatPram?

awgu replied:
This is not important for this PR, but when we expose original parameters, we want to ignore the original parameters that have already been flattened.

`not _is_flattened(param)` correctly captures the logical intention of this method `_get_orig_params()`. We only want this method to return the parameters that have not already been considered by FSDP previously (in some previous FSDP constructor). We previously used `isisnstance(param, FlatParameter)` to determine this because we had an invariant that after an FSDP constructor, all original parameters that have been flattened are de-registered as parameters and only their corresponding `FlatParameter`s are registered.

rohan-varma reviewed (commented):
Main question is whether we really need `_is_flattened` and set_flattened, as it seems that both these APIs and the change isinstance -> type(FlatParam) are accomplishing the same things.

Commit:
Update on "[FSDP] Add `_set_flattened()`; `_is_flattened()`"


For both exposing the original parameters and for TP integration, we cannot only rely on `isinstance(param, FlatParameter)` to ignore already-flattened parameters in `.named_parameters()`. As a simple workaround, we can mark original parameters or `ShardedTensor`s with an attribute `_fsdp_flattened` (saved as a string variable `FSDP_FLATTENED`) to indicate that the parameter/tensor has already been flattened. This issue only arises for recursive/nested FSDP wrapping.

This PR also changes `isinstance(param, FlatParameter)` checks to `type(param) is FlatParameter` because all tensor subclasses that have `_is_param == True` will return `True` for `isinstance(param, <any subclass with _is_param == True>)`. This means that a `ShardedTensor` parameter will return `True` for `isinstance(st, FlatParameter)`, which is not what we want.
https://github.com/pytorch/pytorch/blob/5271494ef21ae0140755a41f3b16a8bd745642b6/torch/nn/parameter.py#L8-L10

[ghstack-poisoned]

awgu commented:
@pytorchbot rebase -s

Commit:
Update on "[FSDP] Add `_set_flattened()`; `_is_flattened()`"


For both exposing the original parameters and for TP integration, we cannot only rely on `isinstance(param, FlatParameter)` to ignore already-flattened parameters in `.named_parameters()`. As a simple workaround, we can mark original parameters or `ShardedTensor`s with an attribute `_fsdp_flattened` (saved as a string variable `FSDP_FLATTENED`) to indicate that the parameter/tensor has already been flattened. This issue only arises for recursive/nested FSDP wrapping.

This PR also changes `isinstance(param, FlatParameter)` checks to `type(param) is FlatParameter` because all tensor subclasses that have `_is_param == True` will return `True` for `isinstance(param, <any subclass with _is_param == True>)`. This means that a `ShardedTensor` parameter will return `True` for `isinstance(st, FlatParameter)`, which is not what we want.
https://github.com/pytorch/pytorch/blob/5271494ef21ae0140755a41f3b16a8bd745642b6/torch/nn/parameter.py#L8-L10

[ghstack-poisoned]
